File: python/61-70/Unique Paths II.py
    # @param obstacleGrid, a list of lists of integers
    # @return an integer
    def uniquePathsWithObstacles(self, obstacleGrid):
        m,n=len(obstacleGrid),len(obstacleGrid[0])
        record=[[0 for j in range (0,n)]  for i in range(0,m)]
        for i in range(0,m):
            if obstacleGrid[i][0]==1:
                break
            record[i][0]=1
        
        for j in range(0,n):
            if obstacleGrid[0][j]==1:
                break
            record[0][j]=1
        
        for i in range(1,m):
            for j in range(1,n):
                if obstacleGrid [i][j]==1:
                    record[i][j]=0
                    continue
                record[i][j]=record[i-1][j]+record[i][j-1]
        return record[m-1][n-1]

File: Cpp/61-70/UniquePathsII.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        if(obstacleGrid.size()==0|| obstacleGrid[0].size()==0)
          return 0;
        int m = obstacleGrid.size();
        int n = obstacleGrid[0].size();
        vector<vector<int> > vec(m,vector<int>(n,0));
        for(int i=0; i<m;i++){
          if(obstacleGrid[i][0]==0)
            vec[i][0] =1;
          else break;
        }
        for(int i=0; i<n;i++){
          if(obstacleGrid[0][i]==0)
            vec[0][i] =1;
          else break;
        }
        for(int i=1; i<m; i++)
          for(int j=1; j<n; j++)
            vec[i][j]= obstacleGrid[i][j]==1?0:vec[i-1][j]+vec[i][j-1];
        return vec[m-1][n-1];
      }
};
